### ISSUE TITLE
ERC4626 does not check the amount of tokens received during the deposit

### ISSUE THREAD

**Issue by vladyan18** [OpenZeppelin/openzeppelin-contracts#4496]
The deposit method assumes that as a result of transferFrom, the pool will receive exactly the number of tokens it expects.

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/d6b63a48ba440ad8d551383697db6e5b0ef84137/contracts/token/ERC20/extensions/ERC4626.sol#L251-L252

However, there are implementations of tokens that may violate this assumption (for example, tokens with transfer fees). In this case, the vault may end up in an inconsistent state.

If this is expected behavior, based on the ERC20 standard, for example, then it might be worth mentioning such requirements for underlying assets.

**Comment by Amxx** [comment]
Yes, this is expected behavior. It would be worth at least mentionning it. 

For the wrapper it really is a strong requirement IMO, but it would also be interresting to try an build a vault that support such tokens.

**Comment by vladyan18** [comment]
> For the wrapper it really is a strong requirement IMO

If I understand correctly, in a wrapper, the user can also get more wrapper token than it should be. Then there will eventually be problems with the burn.

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/d6b63a48ba440ad8d551383697db6e5b0ef84137/contracts/token/ERC20/extensions/ERC20Wrapper.sol#L60-L61


> but it would also be interresting to try an build a vault that support such tokens

Usually, balances before and after the transfer are compared for this. But then there may be a problem with potential reentrancy...

**Comment by frangio** [comment]
Is there any write-up on best practices for DeFi protocols on handling these kinds of tokens? Should they even be handled?

**Comment by vladyan18** [comment]
> Should they even be handled?

Many protocols, such as UniswapV3, [deliberately refuse to support such tokens](https://docs.uniswap.org/concepts/protocol/integration-issues).  But in the case of a UniswapV3, the number of received tokens is still controlled.

Some, like UniswapV2 or Algebra, [have separate support for these tokens](https://github.com/Uniswap/v2-periphery/blob/0335e8f7e1bd1e8d8329fd300aea2ef2f36dd19f/contracts/UniswapV2Router02.sol#L172).

I think that these types of tokens are not very actively used now.

**Comment by vladyan18** [comment]
> Is there any write-up on best practices for DeFi protocols on handling these kinds of tokens?

I don't recall such a document or post, but here's a mention on the OpenZeppelin blog: [Tokens that take fees on transfers](https://blog.openzeppelin.com/workshop-recap-secure-development-workshop-1)
